assets/javascripts/discourse/components/upcoming-events-list.gjs

```diff
@@ -108,17 +108,41 @@ export default class UpcomingEventsList extends Component {
 
   groupByMonthAndDay(data) {
     return data.reduce((result, item) => {
-      const date = new Date(item.starts_at);
-      const year = date.getFullYear();
-      const month = date.getMonth() + 1;
-      const day = date.getDate();
-
-      const monthKey = `${year}-${month}`;
-
-      result[monthKey] = result[monthKey] ?? {};
-      result[monthKey][day] = result[monthKey][day] ?? [];
-
-      result[monthKey][day].push(item);
+      const startDate = moment(item.starts_at);
+      const endDate = item.ends_at ? moment(item.ends_at) : null;
+
+      if (endDate && !startDate.isSame(endDate)) {
+        while (startDate.isSameOrBefore(endDate, "day")) {
+          const year = startDate.year();
+          const month = startDate.month() + 1;
+          const day = startDate.date();
+          const monthKey = `${year}-${month}`;
+          const today = moment();
+
+          // For current events, only push upcoming days
+          if (startDate.isAfter(today)) {
+            result[monthKey] = result[monthKey] || {};
+            result[monthKey][day] = result[monthKey][day] || [];
+
+            result[monthKey][day].push(item);
+          }
```

Comment:
This part is essentially repeated here and on the following else block. You could create a helper function to DRY things up, and just keep the if statement here.

Something like:

```javascript
function addToResult(date, item, result) {
  const year = date.year();
  const month = date.month() + 1;
  const day = date.date();
  const monthKey = `${year}-${month}`;

  result[monthKey] = result[monthKey] || {};
  result[monthKey][day] = result[monthKey][day] || [];
  result[monthKey][day].push(item);
}

//....
// then you can do something like this:

return data.reduce((result, item) => {
  const startDate = moment(item.starts_at);
  const endDate = item.ends_at ? moment(item.ends_at) : null;

  if (!endDate || startDate.isSame(endDate, "day")) {
    addToResult(startDate, item, result);
    return result;
  }

  while (startDate.isSameOrBefore(endDate, "day")) {
    if (startDate.isAfter(today)) {
      addToResult(startDate, item, result);
    }
    startDate.add(1, "day");
  }

  return result;
}, [])
```